src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/NetStandardShims.cs

```diff
@@ -21,6 +35,96 @@ internal static unsafe int GetBytes(this Encoding encoding, ReadOnlySpan<char> s
                 return encoding.GetBytes(pStr, str.Length, pDestination, destination.Length);
             }
         }
+
+        internal static void ReadExactly(this System.IO.Stream stream, Span<byte> buffer) =>
+            ReadAtLeast(stream, buffer, buffer.Length, throwOnEndOfStream: true);
+
+        internal static int ReadAtLeast(
+            this System.IO.Stream stream,
+            Span<byte> buffer,
+            int minimumBytes,
+            bool throwOnEndOfStream = true)
+        {
+            if (minimumBytes > buffer.Length)
+                throw new ArgumentOutOfRangeException(nameof(minimumBytes));
+
+            byte[] rented = CryptoPool.Rent(Math.Min(minimumBytes, 32768));
+            int max = 0;
+            int spaceRemaining = buffer.Length;
+            int totalRead = 0;
+
+            while (totalRead < minimumBytes)
+            {
+                int read = stream.Read(rented, 0, Math.Min(spaceRemaining, rented.Length));
+                max = Math.Max(read, max);
+
+                if (read == 0)
+                {
+                    CryptoPool.Return(rented, max);
+
+                    if (throwOnEndOfStream)
+                    {
+                        throw new System.IO.EndOfStreamException();
+                    }
+
+                    return totalRead;
+                }
+
+                spaceRemaining -= read;
+                totalRead += read;
+                rented.AsSpan(0, read).CopyTo(buffer);
+                buffer = buffer.Slice(read);
+            }
+
+            CryptoPool.Return(rented, max);
+            return totalRead;
+        }
+
+        internal static void AppendData(this IncrementalHash hash, ReadOnlySpan<byte> data)
+        {
+            byte[] rented = CryptoPool.Rent(data.Length);
+
+            try
+            {
+                data.CopyTo(rented);
+                hash.AppendData(rented, 0, data.Length);
+            }
+            finally
+            {
+                CryptoPool.Return(rented, data.Length);
+            }
+        }
+
+#if NETSTANDARD2_0
```

Comment:
Why does this need a compiler def if the whole file is only for downlevel targets?

Reply:
Were there a ns2.1 build it wouldn't be needed... and in the netfx build it's not needed, as the entirety of the IncrementalHash class is provided for compat and already includes the function.

But the `#if` does make things uglier, so I've gone ahead and removed it.